tests/unit/test_usdb_id_file.py

```diff
@@ -0,0 +1,158 @@
+                "http://usdb.animux.de/index.phid=118"
+            ),
+        ),
+        (
+            "broken_usdb_link.url",
+            UsdbIdFileNoParametersMalformedUrlFormatError(
+                "http://usdb.animux.de/index.phid=118"
+            ),
+        ),
+        (
+            "broken_usdb_link.webloc",
+            UsdbIdFileNoParametersMalformedUrlFormatError(
+                "http://usdb.animux.de/index.phid=118"
+            ),
+        ),
+        ("dublicate_url_key.desktop", UsdbIdFileMissingOrDublicateOptionFormatError()),
+        ("dublicate_url_key.url", UsdbIdFileMissingOrDublicateOptionFormatError()),
+        ("dublicate_url_key.webloc", UsdbIdFileMultipleUrlsFormatError()),
+        ("empty.desktop", UsdbIdFileEmptyFileError()),
+        ("empty.url", UsdbIdFileEmptyFileError()),
+        ("empty.webloc", UsdbIdFileEmptyFileError()),
+        ("missing_url_key.desktop", UsdbIdFileMissingKeyFormatError("URL")),
+        ("missing_url_key.url", UsdbIdFileMissingKeyFormatError("URL")),
+        ("missing_url_key.webloc", UsdbIdFileMissingUrlTagFormatError("string")),
+        ("wrong_middle_level.webloc", UsdbIdFileMissingTagFormatError("dict")),
+        (
+            "wrong_top_level.desktop",
+            UsdbIdFileMissingSectionFormatError("Desktop Entry"),
+        ),
+        (
+            "wrong_top_level.url",
+            UsdbIdFileMissingSectionFormatError("InternetShortcut"),
+        ),
+        ("wrong_top_level.webloc", UsdbIdFileMissingTagFormatError("plist")),
+        (
+            "youtube.desktop",
+            UsdbIdFileInvalidDomainMalformedUrlFormatError(
+                "https://www.youtube.com/watch?v=IkLpSgnEqw4", "www.youtube.com"
+            ),
+        ),
+        (
+            "youtube.url",
+            UsdbIdFileInvalidDomainMalformedUrlFormatError(
+                "https://youtu.be/StZcUAPRRac", "youtu.be"
+            ),
+        ),
+        (
+            "youtube.webloc",
+            UsdbIdFileInvalidDomainMalformedUrlFormatError(
+                "https://www.youtube.com/watch?v=IkLpSgnEqw4", "www.youtube.com"
+            ),
+        ),
+        ("empty.usdb_ids", UsdbIdFileEmptyFileError()),
+        ("ids_and_other_stuff.usdb_ids", UsdbIdFileInvalidUsdbIdError()),
+        ("multi-column.usdb_ids", UsdbIdFileInvalidUsdbIdError()),
+        ("broken.json", UsdbIdFileInvalidJsonError()),
+        ("empty.json", UsdbIdFileEmptyFileError()),
+        ("empty_array.json", UsdbIdFileEmptyJsonArrayError()),
+        ("no_array.json", UsdbIdFileNoJsonArrayError()),
+        ("missing_id_key.json", UsdbIdFileMissingKeyFormatError("id")),
+    ],
+)
+def test_invalid_song_id_imports_from_files(
+    resource_dir: str, file: str, expected_error: UsdbIdFileError
+) -> None:
+    path = os.path.join(resource_dir, "import", file)
+    id_file = UsdbIdFile.parse(path)
+    assert repr(id_file.error) == repr(expected_error), f"wrong error from {file}"
```

Comment:
It's a pity you can't just write `id_file.error == expected_error`. `attrs` actually implements `__eq__`, but I guess that doesn't help here because of the fields defined in `Exception`.

Reply:
I think i did so in order to make sure it the exact subclass. Some share the same arguments and so i thought they would therefore match on `__eq__`

Reply:
checked it out... an `==` failes on `expected_error` is Type `UsdbIdFileError`. So any error would be compared against an `UsdbIdFileError` if done with `==`. I thought due to the usage of `attrs.define`, `repr()` should be save

Reply:
Sorry, if that was unclear. I was just commenting on the issue of comparing exceptions. It was not a request for change.
You could write a custom comparison function, but since this is only a test, I think the `repr` hack is very acceptable.

Reply:
Thanks, in this case I would also prefer to leave it like this, as it is currently working and changes might introduce new issues.